services/web/project/app/views.py

```diff
@@ -35,48 +41,71 @@ def new_question():
         q = Question(title, body, current_user.id)
         db.session.add(q)
         db.session.commit()
-        return redirect(url_for("views.question", question_id=q.id))
+        return redirect(url_for(
+            "views.question",
+            method="getQuestion",
+            arg1=q.id))
 
     return render_template("new_question.html", form=form)
 
 
-@views.route('/questions/<question_id>', methods=['GET', 'POST'])
-def question(question_id):
-    # Get question from DB
-    question = Question.query.get(question_id)
-    # Validate that question exists; if not, route to questions forum
-    if question is None:
-        questions()
-        questions()
-        questions()
-    # Get list of answers for question
-    question.answers = Answer.query.filter_by(questionId=question.id).all()
-    question.numAnswers = len(question.answers)
-    # Get the question's creator and assign it as an attribute
-    question.creator = User.query.get(question.userId)
-
-    # Process form
-    form = NewAnswerForm()
-    if form.validate_on_submit():
-        body = form.body.data
-        # Add answer to DB
-        a = Answer(body, current_user.id, question.id)
-        db.session.add(a)
-        db.session.commit()
-        return redirect(url_for('views.question', question_id=question_id))
+@views.route('/questions/', defaults={
```

Comment:
I would not go this way, if you have a check for different methods on the url those should be different methods.Clearer and easier to test later. Single responsibility as much as we can.

Reply:
I disagree on some statements.
I **agree** on:
- renaming `view.py` into `router.py`
- pushing all business login into separate classes like `question_controller.py`

I **disagree** on:
- merging all functions in the controller into one big functions with variable parameters and method parameters. My understanding is that we should stick to `/questions/<q-id>/best-answer/<id>`
- just look at the indentation level of the control statements. What if we need to add two more methods, we should add more `if` statements? Just think of someone else who is not familiar with your code and he/she needs to delve into that complexity comparing to this line https://github.com/NanoProd/StackOverflowMock/blob/60ee103990fe02b90c6a716b4cf0844b93bee3cb/services/web/project/app/views.py#L27
```python
 if(method == "acceptAnswer"): # 1
        result = QuestionCtrl.acceptAnswer(arg1, arg2)
        if(result[0] == "REDIRECT"): # 2 <--------- why not name it something descriptive?
            return redirect(url_for(
                result[1],
                method=result[2],
                arg1=result[3])) # <--------- why not name it something descriptive? what is arg1 in this case?
        else:
            return result
    # Get Queston method
    elif(method == "getQuestion"): # 3
        result = QuestionCtrl.getQuestion(arg1)
        if(result[0] == "REDIRECT_1"): # 4 <--------- why not name it something descriptive? what is the result?
            return redirect(url_for(result[1]))
        elif(result[0] == "REDIRECT_2"): # <--------- why not name it something descriptive? same
            return redirect(url_for(
                result[1],
                method=result[2],
                arg1=result[3]))
        else:
            return render_template(
                result[1],
                question=result[2],
                form=result[3])
    else:
        return redirect(url_for("views.questions"))
```
- SCRUM/Agile they all rely on the flexibility, there is no flexibility in this code, diagrams are not set in stone requirements, if they do not represent what we have they can be redone *send me the source and I will adapt it*

If you are not convinced we can quickly vote as a team at least 3 members to settle this issue, this should be decided one to one :)

Reply:
I don't disagree with you. What I'm trying to get away from if precisely that nested nightmare that results from nested control statements. This situation results from bad design in the first place. We should have established something like url="/controller/methods/param1/param2/.../paramN" and route the request to the relevant controller, call the relevant method and pass it the relevant parameters. This is what I was trying to here, rather unsuccessfully as you pointed out. This because I can't render views or redirect within the controller classes. The methods need to return something to the calling method in route.py.

I'd like to talk it out in person. I feel this issue is more complex than it looks like and its hard to have a discussion by text.